AK/BinaryHeap.h

```diff
@@ -6,108 +6,176 @@
 
 #pragma once
 
-#include <AK/Types.h>
+#include <AK/Noncopyable.h>
+#include <AK/Vector.h>
 
 namespace AK {
 
-template<typename K, typename V, size_t Capacity>
-class BinaryHeap {
+template<typename Node, typename Comparator, typename IndexSetter, size_t inline_capacity = 0>
+class IntrusiveBinaryHeap {
+    AK_MAKE_DEFAULT_COPYABLE(IntrusiveBinaryHeap);
+    AK_MAKE_DEFAULT_MOVABLE(IntrusiveBinaryHeap);
+
 public:
-    BinaryHeap() = default;
-    ~BinaryHeap() = default;
+    IntrusiveBinaryHeap() = default;
 
-    // This constructor allows for O(n) construction of the heap (instead of O(nlogn) for repeated insertions)
-    BinaryHeap(K keys[], V values[], size_t size)
+    IntrusiveBinaryHeap(Vector<Node, inline_capacity>&& nodes)
+        : m_nodes(move(nodes))
     {
-        VERIFY(size <= Capacity);
-        m_size = size;
-        for (size_t i = 0; i < size; i++) {
-            m_elements[i].key = keys[i];
-            m_elements[i].value = values[i];
-        }
-
-        for (ssize_t i = size / 2; i >= 0; i--) {
+        for (ssize_t i = m_nodes.size() / 2; i--;)
```

Comment:
I'm fairly certain `heapify_down` for `size / 2` can be skipped even though BinaryHeap didn't do that before.